### PR TITLE
Fast and correct one-shot systems with a system registry resource

### LINKED ISSUE #2192: One-shot systems via Commands for scripting-like logic
## What problem does this solve or what need does it fill?

Some operations are only done very rarely:  scripting-like game logic that modifies levels, unlocks achievements or handles a boss's special attack come to mind.

These systems clutter the schedule, waste resources as they constantly spin, and are harder to extend.

## What solution would you like?

There are two parts to my proposed solution:

1. Add  `Commands::run_system(system: impl Into<SystemDescriptor>)`. This runs the specified system on the `World` the next time commands are processed.
2. Create a pattern where we store references to system functions on components, and dynamically run those systems when the appropriate trigger is met, allowing for scripting-like flexibility with low-overhead and seamless ergonomic Bevy syntax. 

## What alternative(s) have you considered?

Writing custom commands types each time will be onerous and have much increased boilerplate, and because they take exclusive access to `World`, force the users to learn a new syntax and reduce our ability to parallelize commands of this sort in the future.

This same functionality could also be framed as "dynamically inserting systems", but that may complicate the API somewhat when you often don't end up caring a lot about overhead.

Other approaches to reactivity may be able to fill this niche in some form instead.

## Additional context

Making sure the ordering on this is correct enough is going to be one of the larger challenges. May be better to wait until after #1375. Chaining also needs some careful thought.

Brought up in response to the following [user request](https://discord.com/channels/691052431525675048/742884593551802431/843675331131080704):

>  Is it possible / practical / good practice to pass a function around inside an event? My thought process is an NPC could have an interaction with a very specific procedure following it like opening a secret door across the map or unlocking a new dialogue option with another NPC so much so that having a system that contains every one of these niche interactions inside a match would be rather hideous.

Depending on the efficiency and ergonomics of this, could also be used to support push data-flow for UI (see #254).

Related to #1273.

### PR DESCRIPTION
**Pull request by alice-i-cecile**
# Objective

- Working with exclusive world access is not always easy: in many cases, a standard system or three is more ergonomic to write, and more modularly maintainable.
- For small, one-off tasks (commonly handled with scripting), running an event-reader system incurs a small but flat overhead cost and muddies the schedule.
- Certain forms of logic (e.g. turn-based games) want very fine-grained linear and/or branching control over logic.
- `SystemState` is not automatically cached, and so performance can suffer and change detection breaks.
- Fixes #2192.
- Partial workaround for #279.

## Solution

- Adds a `SystemRegistry` resource to the `World`, which stores initialized systems keyed by their `TypeId`.
- Allows users to call `world.run_system(my_system)` and `commands.run_system(my_system)`, without re-initializing or losing state (essential for change detection).
- Allow users to run systems based on their `TypeId`, enabling more complex user-made abstractions
- Add a `Callback` type to enable convenient use of dynamic one shot systems and reduce the mental overhead of working with `Box<dyn SystemLabel>`.

## Status

- [x] Get sign-off from @cart on this direction, following previous discussions.
- [x] Migrate to a "run all labels" approach: users pass in something that impls Into<SystemLabel> and all all labels are run in sequential insertion order 
- [x] Update this PR to account for changes in #4115.
- [x] Verify tests pass and are complete
- [x] Add an example
- [ ] Simplify example
- [ ] Adapt to the changes in the labelling system, ideally waiting on #5377.

## Future work

- I could not figure out a clean way to implement system chaining, even though it would be useful. Storing heterogenous types is rather nasty.
- Exclusive systems do not work, for the same reason. This should be easily resolved once #4166 is merged.
- Add benchmarks to see how this compares to ordinary schedules (and enable future optimization)
- Enable ordering constraints and parallel execution of system graphs. Probably better to tackle with #4391 in mind.
- Use this API to store all systems, regardless of whether or not they belong to a schedule, and solve #2777.
- Enable recursive `run_system` usage, likely by moving `SystemRegistry` into its own field on `World`.

## Prior attempts

- #2234
- #2417

### CODE REVIEW

File: crates/bevy_core/src/lib.rs [hunk 1]
```diff
@@ -50,6 +50,7 @@ impl Plugin for CorePlugin {
 
         app.init_resource::<Time>()
             .init_resource::<FixedTimesteps>()
+            .init_resource::<SystemRegistry>()
```
Comment (bjorn3) [inline crates/bevy_core/src/lib.rs:53]: This resource is required by bevy_ecs itself. I think it should either be registered directly in `World::new` or stored as it's own field of `World`.

Comment (alice-i-cecile) [reply crates/bevy_core/src/lib.rs:53]: This sort of thing is why I want #3786. Sounds good to me; I think the former solution is less messy.

File: crates/bevy_ecs/src/system/system_registry.rs [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        world: &mut World,
+        system: S,
+    ) {
+        let type_id = TypeId::of::<S>();
+        if self.type_id_registered(&type_id) {
+            self.run_system_by_type_id(world, type_id);
+        } else {
+            self.register_system(world, system);
+            self.run_system_by_type_id(world, type_id);
+        }
```
Comment (MiniaczQ) [inline crates/bevy_ecs/src/system/system_registry.rs:70]: Is there a good reason why an unregistered system should be allowed (through implicit registration) to run?

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_registry.rs:70]: Yes: it makes the API dramatically simpler to use.

With implicit registration, the user can call `world.run_system(my_system)` on an ad-hoc basis anywhere in their code, which is fantastic for prototyping and scripting.

Without it, they have to remember to register it, and then remove that registration when the code changes.


Manual registration is *only* there to enable `TypeId` driven workflows. These are much easier to store and work with than an instance of the function, but I would expect manual registration to be hidden in external user-facing code as well :)

Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_registry.rs:70]: My original argument against this was that users *could* accidentally run arbitrary systems, which would lead to weird behavior, but the ease of use definitely beats that.
On the second thought, centralizing which functions can be run ad hoc would be helpful for the editor in far future.
Going with that idea, maybe throwing warnings would be a decent middle ground? That way users can still prototype easily, but registration would be enforced.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_registry.rs:70]: I don't think the warnings are helpful enough to warrant the level of noise. I'm not sure we're protecting against a real class of errors: this isn't any different than just adding systems to a schedule.

Comment (MiniaczQ) [reply crates/bevy_ecs/src/system/system_registry.rs:70]: Yeah but all schedule systems are known at app launch,
from what I understand the one-shot systems can be registered at any point in runtime.
So in editors case it won't be aware of this until it is ran at least once, right?

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_registry.rs:70]: > Yeah but all schedule systems are known at app launch,

This is only true for now. See #279. I suppose we could opt-in to warnings, but IMO that sort of design is much more easily explored once we *have* a way to display all of the systems in the schedule.

**Review by cart — COMMENTED** [review]
This does seem like a reasonable design. Some assorted thoughts:
* When I pitched my idea for this, it involved "run by label", as that is how we encourage people to refer to systems. Encouraging users to think in "type ids" is a "new" and inconsistent way to reference systems. Given that (as of today) we now auto-label system functions, using actual labels would be better i think (and equally ergonomic).
* I think we should consider embracing the "one label to many systems" design, as that maps best to the label mental model and enables new and interesting call patterns (ex: "run all event handler systems now"). It would require decoupling "registering systems" from "running systems", but imo this would be an improvement. The current implicit "overwrite systems with the same type id" approach feels a bit footgunny. And "pre-defining" systems is already how we do things.

Ultimately I think this pattern could also be embraced by the scheduler (ex: expose a way to queue a run-by-label for the current schedule, using the schedule as the source of systems instead of a SystemRegistry resource  ... or merging the two concepts).  SystemRegistry is really just a very simple schedule impl. In a "running many labels" scenario, it would be cool to (optionally) do that using "parallel system scheduling".

**Comment by alice-i-cecile** [comment]
> When I pitched my idea for this, it involved "run by label", as that is how we encourage people to refer to systems. Encouraging users to think in "type ids" is a "new" and inconsistent way to reference systems. Given that (as of today) we now auto-label system functions, using actual labels would be better i think (and equally ergonomic).

Agreed :) This was not the case when I made this PR, but I'm happy to swap over.

> I think we should consider embracing the "one label to many systems" design, as that maps best to the label mental model and enables new and interesting call patterns (ex: "run all event handler systems now"). It would require decoupling "registering systems" from "running systems", but imo this would be an improvement. The current implicit "overwrite systems with the same type id" approach feels a bit footgunny. And "pre-defining" systems is already how we do things.

Yep, this lines up very nicely with some of the ideas in Stageless about "subgraphs" and "multiple schedules".

> Ultimately I think this pattern could also be embraced by the scheduler (ex: expose a way to queue a run-by-label for the current schedule, using the schedule as the source of systems instead of a SystemRegistry resource ... or merging the two concepts). SystemRegistry is really just a very simple schedule impl. In a "running many labels" scenario, it would be cool to (optionally) do that using "parallel system scheduling".

Yep, this was @maniwani's thinking too: eventually I'd like to converge this.

File: crates/bevy_app/src/system_registry.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        &mut self,
+        system: S,
+    ) -> &mut Self {
+        self.world.register_system(system);
+        self
+    }
+
+    /// Register system a system with any number of [`SystemLabel`]s.
+    ///
+    /// Calls the method of the same name on [`SystemRegistry`].
```
Comment (Nilirad) [inline crates/bevy_app/src/system_registry.rs:20]:
```suggestion
    /// Calls the method of the same name on [`SystemRegistry`].
    /// 
    /// [`SystemRegistry`]: bevy_ecs::system::SystemRegistry
```

There was an undefined link. It should also make it pass CI.

**Comment by alice-i-cecile** [comment]
@IceSentry I've added `Callback` as a first-party type now, per your suggestion :) It's a bit more controversial, but I think it should promote ecosystem compatibility and really lowers the barrier to entry to some fancier tricks (that `PartialEq` implementation was not trivial for example).

File: crates/bevy_ecs/src/system/system_registry.rs [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+            Box::new(IntoSystem::into_system(system));
+        let automatic_system_label: SystemTypeIdLabel<S> = SystemTypeIdLabel::new();
+
+        // This avoids nasty surprising behavior in case systems are registered twice
+        if !self.is_label_registered(automatic_system_label) {
+            self.register_boxed_system_with_labels(
+                world,
+                boxed_system,
+                vec![Box::new(automatic_system_label)],
+            );
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/system/system_registry.rs:119]:
```suggestion
        let automatic_system_label: SystemTypeIdLabel<S> = SystemTypeIdLabel::new();

        // This avoids nasty surprising behavior in case systems are registered twice
        if !self.is_label_registered(automatic_system_label) {
        	let boxed_system: Box<dyn System<In = (), Out = ()>> =
            	Box::new(IntoSystem::into_system(system));
            self.register_boxed_system_with_labels(
                world,
                boxed_system,
                vec![Box::new(automatic_system_label)],
            );
```
This avoids an Box allocation, if system was already registered.

File: crates/bevy_ecs/src/system/system_registry.rs [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+    /// This is only useful for debugging as an external user of this method.
+    ///
+    /// This can be useful when you have a boxed system or boxed labels,
+    /// as the corresponding traits are not implemented for boxed trait objects
+    /// to avoid indefinite nesting.
+    pub fn register_boxed_system_with_labels(
+        &mut self,
+        world: &mut World,
+        mut boxed_system: Box<dyn System<In = (), Out = ()>>,
+        labels: Vec<Box<dyn SystemLabel>>,
```
Comment (MinerSebas) [inline crates/bevy_ecs/src/system/system_registry.rs:171]: Why accept a `Vec` here instead of an `&[]`?
This would allow `register_system` to provide an Array, instead of needing to allocate a Vec.

**Review by MinerSebas — APPROVED** [review]

File: crates/bevy_ecs/src/system/system_registry.rs (hunk 5 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_registry.rs:171]: Good eye! The generic API is a little messy; I'm going to stick to a slice here given how niche this method is.

**Comment by alice-i-cecile** [comment]
@maniwani you've been considering related designs; I'd be curious about your opinion on this implementation if you have the time.

File: examples/ecs/one_shot_systems.rs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+fn setup(mut commands: Commands) {
+    commands
+        .spawn()
+        // The Callback component is defined in bevy_ecs,
+        // but wrapping this (or making your own customized variant) is easy.
+        // Just stored a boxed SystemLabel!
+        .insert(Callback::new(button_pressed))
+        .insert(Triggered);
+    // This entity does not have a Triggered component, so its callback won't run
```
Comment (Nilirad) [inline examples/ecs/one_shot_systems.rs:49]:
```suggestion
    // This entity does not have a Triggered component, so its callback won't run.
```

File: examples/ecs/one_shot_systems.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    println!("A slider was toggled!");
+}
+
+/// Runs the systems associated with each `Callback` component if the entity also has a Triggered component
+///
+/// This could be done in an exclusive system rather than using `Commands` if preferred
+fn evaluate_callbacks(query: Query<&Callback, With<Triggered>>, mut commands: Commands) {
+    for callback in query.iter() {
+        // Because we don't have access to the type information of the callbacks
+        // we have to use the layer of indirection provided by system labels
```
Comment (Nilirad) [inline examples/ecs/one_shot_systems.rs:68]:
```suggestion
        // we have to use the layer of indirection provided by system labels.
```

**Review by Nilirad — CHANGES_REQUESTED** [review]
Some doc-checking. If I have time, I may also review the implementation in the future.

File: examples/ecs/one_shot_systems.rs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+fn button_pressed() {
+    println!("A button was pressed!");
+}
+
+fn slider_toggled() {
+    println!("A slider was toggled!");
+}
+
```
Comment (aevyrie) [inline examples/ecs/one_shot_systems.rs:61]: I found this confusing the first time I read through the example, I though this evaluation system was something special, and callbacks had to be `Component`s with triggers.

I think this could be improved by keeping the example much simpler - like the unit tests you wrote with a simple counter. I finally "got it" after seeing the unit test. This might be a great way to show the on-demand usage of `Callback`s, without the risk of the confusion I experienced. :slightly_smiling_face: 

May I suggest something like this?
```suggestion
/// A resource we will modify with a callback
struct Counter(usize)

/// A bevy system that we will run from a callback instead of inserting it in the schedule
fn increment_counter(counter: ResMut<Counter>) {
    *counter = counter + 1;
}

fn setup(commands: mut Commands) {
    // Callbacks provide a way to run a system as if it was a function call with full access to the ECS.
    commands.insert_resource(Counter(0));
    // Instead of inserting the system into the schedule, we'll invoke it manually using a callback.
    let increment_callback = Callback::new(increment_counter);
    // Recall that commands are queued up and executed at the *end* of the stage.
    commands.run_callback(increment_callback);
}
```

File: crates/bevy_ecs/src/system/system_registry.rs [hunk 9]
```diff
@@ -0,0 +1,607 @@
...
+                .unwrap();
+        });
+    }
+}
+
+/// A struct that stores a boxed [`SystemLabel`], used to cause a [`SystemRegistry`] to run systems.
+///
+/// This might be stored as a component, used as an event, or arranged in a queue stored in a resource.
+/// Unless you need to inspect the list of events or add additional information,
+/// prefer the simpler `commands.run_system` over storing callbacks as events,
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_registry.rs:415]: Not currently. This is desired and should be feasible, but is out of scope for this MVP.

File: crates/bevy_ecs/src/system/function_system.rs [hunk 10]
```diff
@@ -56,8 +56,10 @@ impl SystemMeta {
 // (to avoid the need for unwrapping to retrieve SystemMeta)
 /// Holds on to persistent state required to drive [`SystemParam`] for a [`System`].
 ///
-/// This is a very powerful and convenient tool for working with exclusive world access,
+/// This is a powerful and convenient tool for working with exclusive world access,
 /// allowing you to fetch data from the [`World`] as if you were running a [`System`].
+/// However, simply calling `world::run_system(my_system)` using a [`SystemRegistry`](crate::system::SystemRegistry)
+/// can be significantly simpler and more reliable.
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/function_system.rs:62]:
```suggestion
/// can be significantly simpler and ensures that change detection and command flushing work as expected.
```

File: crates/bevy_ecs/src/system/system_registry.rs [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+
+use crate::schedule::SystemLabel;
+use crate::system::{Command, IntoSystem, System, SystemTypeIdLabel};
+use crate::world::{Mut, World};
+
+/// Stores initialized [`Systems`](crate::system::System), so they can be reused and run in an ad-hoc fashion
+///
+/// Systems are keyed by their [`SystemLabel`]:
+///  - all systems with a given label will be run (in linear registration order) when a given label is run
+///  - repeated calls with the same function type will reuse cached state, including for change detection
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/system_registry.rs:17]: I think this is essential advanced debugging information. Users can insert multiple copies of the same system type, each with their own label, and call them independently.

**Comment by maniwani** [comment]
> @maniwani you've been considering related designs; I'd be curious about your opinion on this implementation if you have the time.

Sorry for the late response. I can describe the setup I have in #4391.

> `@cart`
> When I pitched my idea for this, it involved "run by label", as that is how we encourage people to refer to systems.
>
> I think we should consider embracing the "one label to many systems" design.
>
> Ultimately, I think this pattern could also be embraced by the scheduler. 
> 
> [Running a one-shot system] is really just a very simple schedule impl.

I basically came to the same conclusions as cart. This registry data structure is mentioned in the RFC and included in #4391. The "one label to many systems" design is exactly what system sets have become.

I've taken to just calling it `Systems` (not reflected in my PR yet). All systems and system sets go in the `Systems` resource, and it can generate a dependency graph and pre-allocate an empty `Schedule` for each system set. My impl of `Systems` is designed to let you mutate it while a system or schedule is running.

These are (latest versions of) the important types.

```rust
pub struct Systems {
    // used for lookup
    index: HashMap<SystemLabelId, NodeId>,
    next_id: u64,
    // main storage
    systems: HashMap<NodeId, Option<BoxedSystem>>,
    conditions: HashMap<NodeId, Option<Vec<BoxedRunCondition>>>,
    schedules: HashMap<NodeId, Option<Schedule>>,
    graphs: HashMap<NodeId, DependencyGraph>,
    // needed for graph updates
    hierarchy: DiGraph<NodeId>,
    uninit_nodes: Vec<NodeInfo>,
}

pub enum NodeId {
    System(u64),
    Set(u64),
}

struct DependencyGraph {
    /// dependency graph
    base: DiGraph<NodeId>,
    /// flattened version of the dependency graph (only system nodes and system-system edges)
    flat: DiGraph<NodeId>,
    /// cached topological ordering of `flat`
    topsort: Vec<NodeId>,
    /// `Graph` if the graph (and schedule) need to be rebuilt, `Schedule` if just the schedule, `None` otherwise
    rebuild: Rebuild,
}

pub struct Schedule {
    // All elements are sorted in topological order.
    // NOTE: Might change all the `RefCell` to `Cell<Option<T>>`.
    systems: Vec<RefCell<BoxedSystem>>,
    system_conditions: Vec<RefCell<Vec<BoxedRunCondition>>>,
    set_conditions: Vec<RefCell<Vec<BoxedRunCondition>>>,
    system_ids: Vec<NodeId>,
    set_ids: Vec<NodeId>,
    system_deps: Vec<(usize, Vec<usize>)>,
    sets_of_systems: Vec<FixedBitSet>,
    systems_of_sets: Vec<FixedBitSet>,
}
```

Summarizing: 

- `Systems` is a resource on the `World`, and I've made it so you have to take systems and schedules out of `Systems` (using their corresponding label) in order to run them. Using `resource_scope` would prevent accessing `Systems` again.
- Adding and removing things does not immediately rebuild any schedules. The changes are only propagated up to flag the system sets whose dependency graphs need to be rebuilt (top-sorted again).
- The function to check out a `Schedule` will apply a pending rebuild if the schedule has one, then take the now up-to-date `Schedule` out of its `Option`, move all the required systems and conditions from theirs into it, and return it. 
- All the graph stuff stays with `Systems`, so you can still "remove" systems by label even if they're extracted. You have to return extracted schedules to `Systems` to update them, so if you've removed systems, those will just be dropped at that time.

So basically, you only incur the costs of schedules you use, there's no ownership dilemma, and any `&mut World` context like exclusive systems or commands can extract something from `Systems` and run it.

**Comment by alice-i-cecile** [comment]
FYI if anyone wants to pick this up while I'm on vacation this month feel free! Otherwise I'll wrap this up in September.

**Comment by cart** [comment]
I think we should close this out now that stageless / #6587 is about to land. Running systems on demand will be possible via Schedules:

```rust
app
  .add_system_to(OnDemand, foo)
  .add_system(bar);

fn bar(world: &mut World) {
  world.run_schedule(OnDemand);
} 
```

**Comment by cart** [comment]
(FYI I'm not trying to shutting the conversation down. Happy to open this back up if anyone wants. Stageless doesn't currently include "run systems from commands", but I think that impl would be different (and much smaller) in the new world).

**Comment by cart** [comment]
Also one of the concerns in this PR was that systemstate isn't cacheable, but it is now!

```rust

fn bar(world: &mut World, state: &mut SystemState<Res<Time>>) {
  let delta = state.get(world).delta_seconds();
  if delta_seconds > 2.0 {
      world.run_schedule(OnDemand);
  }
} 
```

**Comment by alice-i-cecile** [comment]
Yep! I agree with this decision!
